### DISCUSSION TITLE
Add `--json` to `borg prune --list`

### DISCUSSION THREAD

**Discussion by WilliamDEdwards** [borgbackup/borg#7021] [Ideas]
I'd like to programmatically extract the archive names that Borg pruned. E.g. with a `--json` switch for `borg prune --list` (like `borg list` has).

`--log-json` is not easily parseable:

```
$ borg prune --log-json --list pen --keep-last 2
{"type": "log_message", "time": 1662974892.256543, "message": "Keeping archive (rule: secondly #1):     mac                                  Mon, 2022-09-12 11:25:59 [aff7f431a4a027ec328437e3d38dd72c86cc81a1ab3099b033d10f1e9e79388a]", "levelname": "INFO", "name": "borg.output.list"}
```

```
>>> import json
>>> line = '{"type": "log_message", "time": 1662974892.256543, "message": "Keeping archive (rule: secondly #1):     mac                                  Mon, 2022-09-12 11:25:59 [aff7f431a4a027ec328437e3d38dd72c86cc81a1ab3099b033d10f1e9e79388a]", "levelname": "INFO", "name": "borg.output.list"}'
>>> line = json.loads(line)
>>> line['message'].split()
['Keeping', 'archive', '(rule:', 'secondly', '#1):', 'mac', 'Mon,', '2022-09-12', '11:25:59', '[aff7f431a4a027ec328437e3d38dd72c86cc81a1ab3099b033d10f1e9e79388a]']
>>> _[5]
'mac'
```

**Comment by dvdheiden** [comment]
It would be really great to have this functionality!

**Comment by ThomasWaldmann** [comment]
What's the point of / use case for extracting the archives that `borg prune` kept?

**Reply by WilliamDEdwards** [reply]
> What's the point of / use case for extracting the archives that `borg prune` kept?

It’s the other way around; I’d like to know what Borg pruned. 

Use case: keeping archives in sync between the real world and a database.

Our customers can create Borg archives via an API. I first add the Borg archive to its database, then do an RPC call to actually create it. 

When Borg prunes, I’d like to know what archives I can delete from the database. I _could_ do that by comparing my database with the archives existing after the prune and removing the difference, but letting Borg do that is more elegant IMO. 

Not keeping archives in the database is not an option, because retrieving them (which is commonly done to e.g. list available backups) would then take at least seconds, due to the overhead of an RPC call plus `borg list`.

**Reply by ThomasWaldmann** [reply]
OK, understood.

**Reply by ThomasWaldmann** [reply]
I created #9222 for this.

**Reply by WilliamDEdwards** [reply]
Thanks!

**Comment by ThomasWaldmann** [comment]
This is implemented in borg 2: `borg prune --json` was added in #9512 and first released in 2.0.0b22.

`--json` prints a single JSON document containing an `archives` list. Each entry has the usual archive keys plus the prune decision:

- `kept`: `true` / `false`
- kept archives additionally get `keep_rule`, `kept_oldest` and `kept_archive_number`
- pruned archives additionally get `deleted_archive_number`

So the archive names that were (or, with `--dry-run`, would be) pruned are directly extractable, no message parsing needed. By default all considered archives are in the output; `--list-pruned` / `--list-kept` narrow it to just one side.

Note this is borg 2 only, it was not backported to 1.4.

Closing as resolved - please open a new discussion or issue if something you need is missing from the output.
